applications: nrf_desktop: Switch from MCUBoot to B0

pdunaj opened the pull request:
When DFU is performed MCUBoot copies application from slot 1
to slot 0. This takes long time during which device is not
operable. Since we don't need any other functionality from
MCUBoot and we need fast app switch time, B0 will be used as
a bootloader.

Jira:DESK-806

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

pdunaj commented:
@wiminordic This should fail on the DFU tests.
First, it uses different image format. Second, version information must be updated in the image before the DFU. Third, I have changed the api of the tool to simplify image selection (you need to select image for the right slot).

pdunaj commented:
Static configuration of PM need to be updated - I will do it tomorrow. At the moment it is not used.

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## checkpatch (informational only, not a failure)

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:1243: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#1243: FILE: applications/nrf_desktop/src/modules/dfu.c:360:
+	if (PM_S1_IMAGE_ID == dfu_slot_id()) {

- total: 0 errors, 1 warnings, 845 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.

```

## pylint issues

```
************* Module dfu
scripts/hid_configurator/modules/dfu.py:302:4: W0702: No exception type(s) specified (bare-except)
scripts/hid_configurator/modules/dfu.py:321:4: W0702: No exception type(s) specified (bare-except)
scripts/hid_configurator/modules/dfu.py:9:0: W0611: Unused imgtool.image imported as img (unused-import)
```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

pdunaj commented:
I think we could start to use zipped files

pdunaj commented on applications/nrf_desktop/configuration/nrf52833_pca10100/b0_ZReleaseB0.conf:
```diff
@@ -0,0 +1,35 @@
+CONFIG_SERIAL=n
+CONFIG_ASSERT=n
+
+CONFIG_PM_PARTITION_SIZE_B0_IMAGE=0x5000
```
Fixed

pdunaj commented on scripts/hid_configurator/modules/dfu.py:
```diff
@@ -6,7 +6,6 @@
 import struct
 import zlib
 import os
-import imgtool.image as img
```
Fixed

pdunaj commented on applications/nrf_desktop/configuration/nrf52833_pca10100/b0_ZDebugB0.conf:
```diff
@@ -0,0 +1,35 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
```
Fixed

pdunaj commented on applications/nrf_desktop/configuration/nrf52833_pca10100/b0_private.pem:
```diff
@@ -0,0 +1,5 @@
+-----BEGIN EC PRIVATE KEY-----
```
Fixed

pdunaj commented on applications/nrf_desktop/CMakeLists.txt:
```diff
@@ -32,6 +32,9 @@ endif()
 set(mcuboot_CONF_FILE
```
I left it on purpose. It is highly possible that we will use MCUBoot for some small devices because of serial recovery dfu.

pdunaj replied:
I added second assert.

pdunaj commented on applications/nrf_desktop/src/modules/dfu.c:
```diff
@@ -347,24 +354,35 @@ static void handle_reboot_request(const struct config_fetch_request_event *event
+	if (info) {
+		u32_t image_size = info->size;
+		u32_t build_num = info->version;
+		u8_t major = 0;
```
I prefer so. To keep the protocol compatible with MCUBoot just in case.

pdunaj commented on scripts/hid_configurator/modules/dfu.py:
```diff
@@ -288,20 +290,38 @@ def is_dfu_image_correct(dfu_image):
-        print('DFU image is invalid')
+    try:
+        img_file = open(dfu_image, 'rb')
+        img_file.seek(0x204)
```
Fixed

pdunaj commented on scripts/hid_configurator/modules/dfu.py:
```diff
@@ -288,20 +290,38 @@ def is_dfu_image_correct(dfu_image):
-    return True
+    img_file.close()
 
+    return magic == 0x8fcebb4c
```
Fixed

pdunaj commented on scripts/hid_configurator/modules/dfu.py:
```diff
@@ -288,20 +290,38 @@ def is_dfu_image_correct(dfu_image):
+def get_dfu_image_version(dfu_image):
+    try:
+        img_file = open(dfu_image, 'rb')
+        img_file.seek(0x214)
```
Fixed

pdunaj commented:
Hi @hakonfam , maybe it would be possible to add an advance option to used different version for S1 then for S0 during the same build? It would simplify testing.

MarekPieta commented on scripts/hid_configurator/modules/dfu.py:
```diff
@@ -286,22 +321,32 @@ def is_dfu_image_correct(dfu_image):
+        print('Invalid image format')
         return False
 
-    return True
```
Do not remove return True here

pdunaj replied:
Thanks!

pdunaj commented on scripts/hid_configurator/modules/dfu.py:
```diff
@@ -26,6 +25,11 @@
 
+UPDATE_IMAGE_MAGIC_COMMON = 0x281ee6de
+UPDATE_IMAGE_MAGIC_FWINFO = 0x8fcebb4c
+UPDATE_IMAGE_MAGIC_COMPATIBILITY = 0x00003402
```
This is true but we only support nrf52 devices. I would not spent time on this one as it is a wider problem.

hakonfam commented:
> Hi @hakonfam , maybe it would be possible to add an advance option to used different version for S1 then for S0 during the same build? It would simplify testing.

As discussed offline, it is possible, but would require 2 out of 5 amount of work. The changes needed are in `zephyr/include/scripts/.../linker.ld`. Some symbol would have to be set based on whether the link is run against S0 or S1, and then this symbol value must be used in fw_info.c as the version. IMO the best solution would be to simply bump the version for S0 if `CONFIG_B0_BUMP_S1_VERSION=y`.

pdunaj commented:
Rebased

pdunaj commented:
@MarekPieta , I am actually wondering if not to leave support for MCUBoot as an option... In FW this should be really simple. It's not really much complicated on the tool side. The good thing would be that we could easily enable mcuboot with serial USB recovery on the small devices.

pdunaj commented:
@wiminordic , could you tell how much effort would it be to update the tests?

pdunaj commented:
CI failure
```
    from ecdsa import SigningKey

ModuleNotFoundError: No module named 'ecdsa'
```

pdunaj commented:
rebased

pdunaj commented:
Hi @wiminordic , I have rebased and added a code to handle zip packages with update images. You can try to work with it now.
There are two disclaimers tough:
- on master package is not created if application uses B0; I have added this but the code will change for sure - @hakonfam knows about the problem
- for the same reason name of the zip file may also change.
Anyway you could check how it works. There is no huge difference - you use zip file instead of the bin file.

pdunaj commented:
We are waiting for https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/2094.
I will rebase once merged.

pdunaj commented:
Rebased on top of the master.

Hi @wiminordic , the zip file is in `build/zephyr/dfu_application.zip`. I think that in my commit it was present directly in the build directory. Can you take the artifact from the new location?

pdunaj commented:
There are some CI failures:
```
[0/274] Creating public key from private key used for signing
Traceback (most recent call last):
  File "/jenkins_workspace/workspace/latest_fw-nrfconnect-nrf_PR-1990/nrf/scripts/bootloader/keygen.py", line 8, in <module>
    from ecdsa import SigningKey
ModuleNotFoundError: No module named 'ecdsa'
[1/274] Preparing syscall dependency handling
[2/274] Creating directories for 'b0_subimage'
FAILED: zephyr/nrf/subsys/bootloader/generated/public.pem 
cd /jenkins_workspace/workspace/latest_fw-nrfconnect-nrf_PR-1990/sanity-out/nrf52840_pca10056/nrf_desktop/test_zdebug_b0/zephyr && /usr/local/bin/python3.6 /jenkins_workspace/workspace/latest_fw-nrfconnect-nrf_PR-1990/nrf/scripts/bootloader/keygen.py --public --in /jenkins_workspace/workspace/latest_fw-nrfconnect-nrf_PR-1990/nrf/applications/nrf_desktop/configuration/nrf52840_pca10056/b0_private.pem --out /jenkins_workspace/workspace/latest_fw-nrfconnect-nrf_PR-1990/sanity-out/nrf52840_pca10056/nrf_desktop/test_zdebug_b0/zephyr/nrf/subsys/bootloader/generated/public.pem
[4/274] Generating misc/generated/syscalls_subdirs.trigger
[5/274] No download step for 'b0_subimage'
```

I do not replicate it locally. Maybe this is the problem? `ModuleNotFoundError: No module named 'ecdsa'`

pdunaj commented:
Hi @hakonfam , I have one commit for missing header (`e9fdf1ede0ffde9b8584562c9471938d3785fd38`) are you find with having it in?

wiminordic commented:
> Rebased on top of the master.
> 
> Hi @wiminordic , the zip file is in `build/zephyr/dfu_application.zip`. I think that in my commit it was present directly in the build directory. Can you take the artifact from the new location?

Yes, that is fine.

pdunaj commented:
@wiminordic , thanks. Do you know what can be wrong with the CI?

hakonfam commented:
> Hi @hakonfam , I have one commit for missing header (`e9fdf1ede0ffde9b8584562c9471938d3785fd38`) are you find with having it in?

Sure, what symbols were missing?

pdunaj commented:
Hi @thst-nordic , thanks for the commit. Unfortunatelly it seems that the problem is still there.

> ModuleNotFoundError: No module named 'ecdsa'

pdunaj commented:
Hi @wiminordic , the CI is green now. Could you tell when we can merge?

wiminordic commented:
@pdunaj Go ahead, merge.

pdunaj commented:
Hi @hakonfam , thanks. `__set_MSP`, `__set_PSP` and several others.
